### PR TITLE
[WIP] EZP-24177: Improve Search to allow Field criterion use on non searchable Fields

### PR DESCRIPTION
**Pull request by alongosz**
> Implements: [EZP-24177](https://jira.ez.no/browse/EZP-24177)
> Status: **Work in progress**

This PR proposes changing meaning of `FieldDefinition::isSearchable` flag. So far, it only meant that a Field is searchable. However, as EZP-24177 describes:

> Search in Platform stack is a replacement both for Search and Fetch functions in Legacy

In order to use `Field` Criterion (and possibly others) to create fetch queries for non-searchable Fields  we need `isSearchable` to mean that a Field is indexed for FullText search and also allow querying non-searchable fields.

**TODO**:
- [x] Change API docs on isSearchable.
- [x] Update SQL FullTextMapper to use `Indexable::getFullTextData`.
- [x] Update ES StandardMapper to use `Indexable::getFullTextData` and `Indexable::getFilterData`.
- [x] Introduce `Indexable::getFilterData` method returning non-fulltext search fields.
- [x] Introduce `Indexable::getFullTextData` method returning array of strings for FullText search.
- [x] Deprecate `FieldType::isSearchable`, `FieldType::getSortInfo`, `FieldValue::$sortKey`
- [x] Added test case testing searching using Criterion\Field on non-searchable fields.
- [ ] Make `FieldDefinition::$isSearchable` not controlled by `FieldType::isSearchable` - *needs decision if shouldn't be disabled in UI at all or should be disabled based on different criteria (note: rather to do in repository-forms than kernel)*.
- [x] Make all Search Engines to use `FieldDefinition::$isSearchable` and `Indexable::getFullTextData` to index full text data.
- [ ] Make Legacy storage field value converter implementations to use `Indexable` to get "sort" data.
- [ ] *TBD*...

### CODE REVIEW

**Comment by andrerom** [comment]
Besides changes needed in SPI/Persistence as well, here is possible change on FieldType to clarify:
```diff
diff --git a/eZ/Publish/SPI/FieldType/FieldType.php b/eZ/Publish/SPI/FieldType/FieldType.php
index f7e57b0..42192fb 100644
--- a/eZ/Publish/SPI/FieldType/FieldType.php
+++ b/eZ/Publish/SPI/FieldType/FieldType.php
@@ -178,7 +178,11 @@ interface FieldType
     public function applyDefaultSettings(&$fieldSettings);
 
     /**
-     * Indicates if the field type supports indexing and sort keys for searching.
+     * Indicates if the field is indexed for FullText search.
+     *
+     * Implies that this fieldType must implement {@see eZ\Publish\SPI\FieldType\Indexable} with a
+     * {@see eZ\Publish\SPI\Search\FieldType\FullTextField} field, if false field can still be queried and sorted
+     * as long as fieldType implements {@see eZ\Publish\SPI\FieldType\Indexable}.
      *
      * @return bool
      */
```


/cc @pspanja

**Comment by pspanja** [comment]
I think eventually we could have both `FieldDefinition::isSearchable` and `FieldDefinition::indexFullText`.
For example, if someone used `TextBlock` to store large amount of text not useful for search, it would unnecessarily inflate index and degrade indexing performance if there was no way to control it.
Similar if we implement indexing `File` content with Apache Tika, one might want to control it for the same reasons.

Given we might not want to make database changes at this point, maybe we should introduce `FieldDefinition::indexFullText` using the same database field, and keep `FieldDefinition::isSearchable` for the future usage as described above, but remove all current use of it.

Does that make sense?

**Comment by andrerom** [comment]
I'm not sure I follow, afaik we have the following needs:
- fieldType hints about support for _(we current kind of duplicate this between type and Indexable)_
 - fulltext search
 - filter / sort _(light)_ index
- user control over if full text search should be done or not per field

**Comment by pspanja** [comment]
What I'm proposing is separate user control if the field data is indexed for:

- `Field` criterion and sort clause / light search -- by `isSearchable` (or better `indexData` or similar)
- full text search -- by `indexFullTextData` (or similar)

My question is: if we eventually want that control, we should use something else from `isSearchable` to control full text data indexing.

Separate from that, we should probably introduce a separate method for full text data in the `Indexable` interface, so we can handle it from the outside and not for each field type separately in `getIndexData()`.

**Comment by pspanja** [comment]
I don't quite get what is the meaning of [`FieldType::isSearchable()`](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/SPI/FieldType/FieldType.php#L183) with this change?

For example `LandingPage` field type is not searchable, meaning it provides no data for indexing.
What happens if you try to search on it with `Field` criterion?

**Comment by alongosz** [comment]
The way I understood one of the goals of the Jira issue:
> Search in Platform stack is a replacement both for Search and Fetch functions in legacy. (...)
> Right now this is disallowed and Field, MapLocationDistance and potentially others meant for querying (filtering like in legacy fetch functions), so the change would be to:
> - allow despite this flag [isSearchable], simplifying the queries needed there

is to allow using Criterions like `Criterion\Field('eztext', ...` as shown in the test case 2595192. Solution I found, in 9bf0d51, is to get, instead of searchable field map, a field map filtered by `$fieldIdentifier`. I'm aware this could be a performance problem, but right now I don't have a better idea.

Anyway, I don't think it's possible to do it in a clean way, because, e.g. for Legacy/SQL, it is still [based on indexColumn](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/Field.php#L116) which in this case is `ezcontentobject_attribute`.`sort_key_string`. Hence the change in d3d57d3. Similar set of changes would be required for Solr and Elastic. With this approach we also can't avoid changing specific fields, so I'm not happy with my solution at all.

@pspanja is there a better way to do it, keeping in mind we want to achieve what's in issue description (quoted above)?

**Comment by pspanja** [comment]
> Anyway, I don't think it's possible to do it in a clean way, because, e.g. for Legacy/SQL, it is still based on indexColumn which in this case is ezcontentobject_attribute.sort_key_string. Hence the change in d3d57d3. Similar set of changes would be required for Solr and Elastic. With this approach we also can't avoid changing specific fields, so I'm not happy with my solution at all.

`TextBlock` field type can store a string that is way to big for `sort_key_string`, so I don't think it make sense to handle it that way. There will be field types that:

1. do not index anything, not in Solr or in Legacy with `sort_key_*`
2. index data in Solr through `Indexable`, but don't index data in Legacy (like `TextBlock`)

I think we should first decide how to handle those cases.

Also, if `FieldDefinition::isSearchable` now means indexing full text data, since `Indexable::getIndexData()` provides both full text data and fields for search/sort with `Field` criterion and sort clause, we will have to control indexing of full text data through that method. This seems awkward and we should probably improve the API there, hence my comment:

>> Separate from that, we should probably introduce a separate method for full text data in the `Indexable` interface, so we can handle it from the outside and not for each field type separately in `getIndexData()`.

**Comment by alongosz** [comment]
The main problem here is that we're trying to implement search on something that is not intended to be indexed & searched. If Search should replace Legacy Fetch, then maybe all fields just should be searchable? Otherwise, as Field mappings for search are strongly tied to Search Engines gateways, this implies may changes for every SE.

So, I also would prefer to handle it from the outside. I'm not sure about `Indexable` interface though. IMHO the best would be adding special handling (a separate interface) for Fields that are not intended for search. `Indexable` interface is implemented on all fields, so this would also imply many changes (possibly to custom FieldTypes as well).

Next, beside handling for field types, we need to decide on where and when to handle it in a generic way. The last possibility is [here](https://github.com/ezsystems/ezpublish-kernel/blob/v6.6.1/eZ/Publish/Core/Repository/SearchService.php#L186) as `$this->searchHandler->findContent` is SearchEngine-specific. 

Approach I'm thinking of would involve catching `InvalidArgumentException` (or preferably some dedicated Exception) and handle special case. Not the cleanest way, but maybe requires less changes and doesn't change the meaning of `isSearchable`.

// cc @pspanja, @andrerom

**Comment by pspanja** [comment]
One idea here would be to stop using `FieldType::isSearchable()` and instead discover how the field type is searchable from `Indexable` implementation for the field type. For the full text indexing we should introduce a separate method `Indexable::getFullTextData()`. This should be usable for all search engines.

`FieldDefinition::$isSearchable` then becomes free from the control of `FieldType::isSearchable()`, user can set it how it wants and it only controls full text indexing. Indexing of Content Field data is controlled by `Indexable::getFieldData()`. We use `getDefaultMatchField()` and `getDefaultSortField()` to discover if the field is searchable or sortable with `Field` criterion/sort clause.

The problem here is Legacy Search Engine, which does not use `Indexable` to index field data in `sort_key_*` columns (it does use it for full text indexing). So we will need to introduce it there.

The problem with Legacy engine is we're limited with schema. One example for this is `TextBlock` field type, which can have a string that is simply too long to be indexed in the `sort_key_string` column. I think we should decide how to handle that per field type, in the field value converter implementation. For example `TextBlock` field type can use only the part of string provided by `Indexable`, `Author` field type can concatenate author name data. However, short of modifying the database schema, there will be differences and they should be documented.

Additionally: with `Indexable` we can have separate default search and sort fields, with legacy we can use only one of those.

### Summary

Deprecate:

- `FieldType::isSearchable()`
- `FieldType::getSortInfo()`
- Persistence `FieldValue::$sortKey`

Introduce:

- `Indexable::getFullTextData()`

Refactor:
- `FieldDefinition::$isSearchable` is not controlled by `FieldType::isSearchable()`
- All search engines to use `FieldDefinition::$isSearchable` and `Indexable::getFullTextData()` to index full text data
- Legacy storage field value converter implementations to use `Indexable` to get "sort" data. It should be possible to do this for our field types without making changes to how data is indexed, after that we can start making more field types searchable.

@andrerom what do you think?

**Comment by andrerom** [comment]
> Deprecate FieldType::getSortInfo()

This is actually not on the SPI interface, so should be fine. But then I assume we are safe that the data is in same or still valid format for legacy storage engine on indexer class?

> FieldDefinition::$isSearchable is not controlled by FieldType::isSearchable()

true, but ideally it should a still only be allowed to be set to true if field type has indexer cable of fulltext indexing. 


> The problem with Legacy engine is we're limited with schema. 

I think this is something the engine needs to take care about, limiting length on word boundaries.

> @andrerom what do you think?

One thing seems to be missing, we would need method on Indexable that replaces `FieldType::isSearchable()` for FieldCriterion/FieldSortClause right? so it can throw exception if field does not support being filtered / sorted.  For FullText it seems we don't need anything, as we don't care about actual fields on when searching by fullText right?

**Comment by gggeek** [comment]
For clarity of exposition and shared understanding: are we talking about 3 separate needs/actions here? If so we should maybe adopt the following terms:

1. 'searching' for content
2. 'filtering' content
3. 'sorting' content

1. is switched on/off per field in the definition of content type, and is the default action carried out by users
2. is always available, and is carried out by code
3. does this apply to the results of 1, 2, or both ?

I am not a great fan of the 'indexable' term, as in databases indexes are uses outside of full-text search

**Comment by pspanja** [comment]
@andrerom 

> One thing seems to be missing, we would need method on Indexable that replaces FieldType::isSearchable() for FieldCriterion/FieldSortClause right? so it can throw exception if field does not support being filtered / sorted.

Actually we have `getDefaultMatchField()` and `getDefaultSortField()` to use for that.

> For FullText it seems we don't need that for the criteria, but we would need it if we want to make sure FieldDef->isSearchable is only allowed to be true if fulltext indexing is supported.

We could also do that by checking what `getFullTextData()` returns.

**Comment by pspanja** [comment]
@gggeek 

> 1 is switched on/off per field in the definition of content type, and is the default action carried out by users

Yes, this is AKA full text search.

> 2 is always available, and is carried out by code

Yes, by code, but not always available, you could still have a field type with [`Unindexed`](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/Unindexed.php) implementation of the `Indexable`. We just move control of it from being hardcoded in the `FieldType` implementation to the `Indexable` implementation, which is easier to override.

> 3 does this apply to the results of 1, 2, or both ?

Both.

> I am not a great fan of the 'indexable' term, as in databases indexes are uses outside of full-text search

True, but with search engines like Elastic and Solr it's quite common term for what we do with it.

**Comment by pspanja** [comment]
Regarding how to control `FieldDefinition::$isSearchable`, yes we can do it by checking `Indexable` implementation, but from my perspective it's more an UI thing and we should rather leave it to be set without any control. For example one could change `Indexable` implementation for the field type, changing it so that field is made not searchable, while the fields in the ContentTypes are still marked as searchable. But it's not big deal IMO.

**Comment by pspanja** [comment]
@andrerom 

> This is actually not on the SPI interface, so should be fine. But then I assume we are safe that the data is in same or still valid format for legacy storage engine on indexer class?

Yes, this should be the case for those field types that are searchable in both engines.

**Comment by andrerom** [comment]
>  For example one could change Indexable implementation for the field type, changing it so that field is made not searchable, while the fields in the ContentTypes are still marked as searchable. But it's not big deal IMO.

yes, but form API perspective on create and update I would like that we either throw if value is set to true but field type does not support it, or that we continue doing like was added recently, just force it to false if not supported. Either way we should not persist true if field type does not support it imho.


> We could also do that by checking what getFullTextData() returns.

yes we can, but this will most likely need spi field and fieldDefinition so not always applicable.

File: eZ/Publish/SPI/Search/FullTextFieldValue.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * The wrapper for \eZ\Publish\SPI\Search\Field of the type \eZ\Publish\SPI\Search\FieldType\FullTextField.
+ */
+class FullTextFieldValue extends Field
+{
+    /**
+     * @param mixed $value
+     */
+    public function __construct($value)
+    {
+        parent::__construct('fulltext', (string) $value, new FieldType\FullTextField());
```
Comment (pspanja) [inline eZ/Publish/SPI/Search/FullTextFieldValue.php]: You mean in `getFulltextData()`? I thought you went with returning an array of strings there?
This basically hardcodes field name, and we might need it for other stuff, like indexing other full text fields (per Content field, then implementing `FULLTEXT` operator for `Field`/`CustomField` criterion).

Also, since this is API (SPI), I would do this kind of thing in the implementation and not here.

Comment (alongosz) [inline eZ/Publish/SPI/Search/FullTextFieldValue.php]: @pspanja: Sorry, I forgot to mention that I had to go back to returning `Search\Field[]` from `getFullTextData`, so it is in sync with `getFilterData` - for mappers to use them in a generic way. 
Usage is shown for ES in f435c88.

It allows me not to repeat 
`new Search\Field('fulltext', $value, new FieldType\FullTextField())` 
for each field (9391fa4).
> Also, since this is API (SPI), I would do this kind of thing in the implementation and not here.

I agree, API/SPI is probably not a good place for this, however I would like to have it somewhere common for all search engines implementations (`eZ\Publish\Core\Search\Common`?).
But if you still feel it's bad, I'll change it back to `Search\Field` :)

**Comment by alongosz** [comment]
PR was rebased to take into account changes from #1849 (without it tests crash).

File: eZ/Publish/SPI/Search/FullTextFieldValue.php (hunk 1 above)
Comment (andrerom) [inline eZ/Publish/SPI/Search/FullTextFieldValue.php]: @alongosz I know I implied that in
```
Ideally we should maybe:
- deprecate `getIndexData` and leave it as-is also returning fulltext data
- add `getFullTextData`, somewhat done
- add `<insert-name>` for non full text, _maybe `getFilterData,  `filter` is not the perfect domain word, but it’s the most applicable I can think of right now._

however to allow Solr/ES code to handle it somewhat uniform, return data for these two new methods should be somewhat in sync, or at least the wrapper `getFields()` should be able to handle the difference in format, so generic code in `mapFields()` can stay generic (e
```

However I also implied it could be mapped in Solr/ES.

Specific example: https://github.com/ezsystems/ezplatform-solr-search-engine/blob/master/lib/FieldMapper/ContentTranslationFieldMapper/BlockDocumentsContentFields.php#L78-Lundefined

instead of having `appendField()` below, the line linked to above can so something like ` $indexFields = $this->geSearchFields($field, $fieldDefinition);`, and the child Fulltext class extending this can override that protected method.


So you two can decide on whatever format is best to return from these two methods as long as `geSearchFields` is able to map it to a common format `mapFields()` can just use ;)

**Comment by alongosz** [comment]
I've rebased it once again to align with changes from #1855.

Continuing discussion from an outdated diff:

> @andrerom
> However I also implied it could be mapped in Solr/ES.
>
> Specific example: https://github.com/ezsystems/ezplatform-solr-search-engine/blob/master/lib/FieldMapper/ContentTranslationFieldMapper/BlockDocumentsContentFields.php#L78-Lundefined
> 
> instead of having `appendField()` below, the line linked to above can so something like ` $indexFields = $this->geSearchFields($field, $fieldDefinition);`, and the child Fulltext class extending this can override that protected method.

> @pspanja
> I would go for returning an array of strings, then it's up to the engine to handle it however it needs to. Eventually we should be able to deprecate `FullTextField` altogether.

Ok :) `getFullTextData` returns `string[]` now.

Specific implementations:
- ES changed in 6bc7053. 
- Solr changed in [PR #82](https://github.com/ezsystems/ezplatform-solr-search-engine/pull/82) / impl. in: [3560004](https://github.com/ezsystems/ezplatform-solr-search-engine/pull/82/commits/35600041d54915a4138cc3cc9d5d77c21d05dea6).
- SQL Search changed in 89896c4.

**Comment by alongosz** [comment]
We've decided to postpone refactoring related to storage engine until we improve the engine itself.

Before I ask you for the final review I need to cleanup this PR removing parts that are not applicable anymore / incomplete:

- Changes to API docs on isSearchable (IMO we don't really established clear direction in which this should be heading).
- Deprecations of `FieldType::getSortInfo`, `FieldValue::$sortKey` (due to postponing refactoring in legacy storage engine). Note: deprecating `FieldType::isSearchable` still makes sense for the scope of this PR.

So basically what this PR is - it refactors `Indexable` splitting `getIndexData` into `getFilterData` and `getFullText`.
**My question is - can I do something more here**? If not, then probably PR title (and resulting merge commit msg) should be changed, because this PR right now does not solve issue, it's rather only first step.

ping @andrerom @pspanja

File: eZ/Publish/Core/FieldType/Price/SearchField.php [hunk 2]
```diff
@@ -8,79 +8,11 @@
...
-use eZ\Publish\SPI\Persistence\Content\Type\FieldDefinition;
-use eZ\Publish\SPI\FieldType\Indexable;
-use eZ\Publish\SPI\Search;
+use eZ\Publish\Core\FieldType\Unindexed;
 
 /**
  * Indexable definition for string field type.
  */
-class SearchField implements Indexable
+class SearchField extends Unindexed
```
Comment (pspanja) [inline eZ/Publish/Core/FieldType/Price/SearchField.php]: No need for this, you should instead remove this file completely and instead register `Unindexed` in the container, as is [already the case for some other field types](https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/settings/indexable_fieldtypes.yml#L148).

Comment (alongosz) [inline eZ/Publish/Core/FieldType/Price/SearchField.php]: Nice, I didn't notice that.
Fixed in cb01e54, thanks.

File: eZ/Publish/Core/Search/Elasticsearch/Content/Mapper/StandardMapper.php [hunk 3]
```diff
@@ -348,6 +329,59 @@ protected function mapFields(Content $content, Type $contentType)
...
+                $filterField->type
+            );
+        }
+
+        if (!$fieldDefinition->isSearchable) {
+            return $searchFields;
+        }
+
+        $fullTextData = $fieldType->getFullTextData($field, $fieldDefinition);
+        foreach ($fullTextData as $fullTextValue) {
```
Comment (pspanja) [inline eZ/Publish/Core/Search/Elasticsearch/Content/Mapper/StandardMapper.php]: `FullText` field is already handled by `MultipleStringField` mapper, so it would be OK to construct a single field passing `$fullTextData` (an array of strings) to it derectly.

File: eZ/Publish/Core/FieldType/DateAndTime/SearchField.php [hunk 4]
```diff
@@ -76,4 +77,32 @@ public function getDefaultSortField()
...
+     * {@inheritdoc}
+     */
+    public function getFullTextData(Field $field, FieldDefinition $fieldDefinition)
+    {
+        $dateTime = new DateTime();
+        $dateTime->setTimestamp($field->value->data['timestamp']);
+
+        return [
+            // format taken from {@see \eZ\Publish\Core\FieldType\DateAndTime\Value::$stringFormat}
+            $dateTime->format('U'),
```
Comment (pspanja) [inline eZ/Publish/Core/FieldType/DateAndTime/SearchField.php]: Seconds since the Unix Epoch? Not sure how usable this is for full text search :)
Also, I'm not sure why this was added? Maybe something for a separate PR.

File: eZ/Publish/Core/FieldType/Date/SearchField.php [hunk 5]
```diff
@@ -83,4 +83,40 @@ public function getDefaultSortField()
...
+    public function getFullTextData(Field $field, FieldDefinition $fieldDefinition)
+    {
+        // The field type stores date value as a timestamp of the start of the day in the
+        // environment's timezone.
+        // We format this as Y-m-d and add Z to signify UTC (zero offset).
+        $dateTime = new DateTime();
+        $dateTime->setTimestamp($field->value->data['timestamp']);
+
+        return [
+            $dateTime->format('Y-m-d\\Z'),
```
Comment (pspanja) [inline eZ/Publish/Core/FieldType/Date/SearchField.php]: This was not done originally, with changes like this we should rather not add in new stuff as well.
*(BTW this also means it could be integration tested, we already have a framework for that in place)*

**Comment by pspanja** [comment]
> We've decided to postpone refactoring related to storage engine until we improve the engine itself.

OK, then we won't control it from field type `Indexable` implementation for now.

> Changes to API docs on isSearchable (IMO we didn't really establish clear direction in which this should be heading).

You mean on the `FieldDefinition` right? I think we agreed `FieldType::$isSearchble` gets deprecated, while `FieldDefinition::$isSearchable` should mean indexing for full text.

**Comment by alongosz** [comment]
> > Changes to API docs on isSearchable (IMO we didn't really establish clear direction in which this should be heading).
>
> You mean on the `FieldDefinition` right? I think we agreed `FieldType::$isSearchble` gets deprecated, while `FieldDefinition::$isSearchable` should mean indexing for full text.

@pspanja I meant changes introduced in e1940c0 (sorry I wasn't specific enough when writing this).
Deprecation of `FieldType::$isSearchable` makes sense and should stay*.

However other changes on API docs seems to be half-true, because right now we have fields that does not return full-text data but are searchable (e.g. `FieldType\Time`, `FieldType\BinaryFile`).

\* while at it, we should also update [this code in repository-forms](https://github.com/ezsystems/repository-forms/blob/v1.5.0-rc1/lib/Form/Type/FieldDefinition/FieldDefinitionType.php#L112). It might be as simple as replacing it with `!$data->fieldDefinition->isSearchable`, but I need to test that ;)

// cc @andrerom

File: eZ/Publish/Core/FieldType/DateAndTime/SearchField.php (hunk 4 above)
Comment (alongosz) [inline eZ/Publish/Core/FieldType/DateAndTime/SearchField.php]: Right, we agreed before not to change current behavior, I must've forgotten about this code. Moreover Epoch for FullText makes no sense :) Fixed it in 06e437d.

File: eZ/Publish/Core/FieldType/Date/SearchField.php (hunk 5 above)
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Date/SearchField.php]: True, fixed in 06e437d. By integration test did you mean test checking that this does not return FullText or something else?

File: eZ/Publish/Core/Search/Elasticsearch/Content/Mapper/StandardMapper.php (hunk 3 above)
Comment (alongosz) [inline eZ/Publish/Core/Search/Elasticsearch/Content/Mapper/StandardMapper.php]: Ok, simplified in a3391ca :)

**Comment by andrerom** [comment]
> while at it, we should also update this code in repository-forms. It might be as simple as replacing it with !$data->fieldDefinition->isSearchable

does not compute does it? the code there is for letting users set `$data->fieldDefinition->isSearchable`, aka if field will be allowed to be full text searchable, so point of disabling it is to check if field can be fulltext indexed at all.

**Comment by alongosz** [comment]
> does not compute does it? the code there is for letting users set `$data->fieldDefinition->isSearchable`, aka if field will be allowed to be full text searchable, so point of disabling it is to check if field can be fulltext indexed at all.

Yes, of course. Sorry, I should've checked it more carefully. 

Then the question remains: how we can refactor this in repository-forms? Looking at previous comments, I'm not sure if we've reached a conclusion on that. 
For instance, if we base this on `Indexable::getFullTextData`, then this will be a BC break, because not all field types with `isSearchable = true` return full-text data.

**Comment by alongosz** [comment]
Closing this for now until we decide on some feasible goal (preferably a number of PRs with small changes, but TBA at this point)
